zaza/openstack/charm_tests/ceph/tests.py

```diff
@@ -522,68 +517,122 @@ def get_local_osd_id(self, unit):
+        self.assertEqual(results['osd-ids'], osd_id)
+        zaza_model.run_on_unit(unit, 'partprobe')
+
+    def purge_remove_one_osd(self, unit):
+        """Remove one device from osd."""
+        block_devs = self.get_osd_devices_on_unit(unit)
+        # Should have more than 1 OSDs to take one out and test.
+        self.assertGreater(len(block_devs), 1)
+
+        # Get complete device details for an OSD.
+        key = list(block_devs)[-1]
+        device = {
+            'osd-id': key if key.startswith('osd.') else 'osd.' + key,
+            'block-device': block_devs[key][0]['devices'][0]
+        }
+
+        self.remove_disk_from_osd_unit(unit, device['osd-id'], is_purge=True)
+        return device
 
     def test_cache_device(self):
         """Test replacing a disk in use."""
         logging.info('Running add-disk action with a caching device')
         mon = next(iter(zaza_model.get_units('ceph-mon'))).entity_id
         osds = [x.entity_id for x in zaza_model.get_units('ceph-osd')]
-        params = []
+        osd_info = dict()
+
+        # Purge Remove one of the two disks.
+        logging.info('Removing single disk from OSDs')
         for unit in osds:
-            loop_dev = self.loop_devs[unit]
-            params.append({'unit': unit, 'device': loop_dev})
+            device_info = self.purge_remove_one_osd(unit)
+            logging.info(
+                "Found device {} on unit {}".format(device_info, unit)
+            )
+            # Save for later use.
+            osd_info[unit] = device_info
+        logging.debug('Removed OSD Info: {}'.format(osd_info))
+        zaza_model.wait_for_application_states()
+
+        logging.info('Recycling previously removed disks')
+        for unit in list(osd_info):
+            block_dev = osd_info[unit]['block-device']
+            logging.info("Found device {} on unit {}".format(block_dev, unit))
+            self.assertNotEqual(block_dev, None)
             action_obj = zaza_model.run_action(
                 unit_name=unit,
                 action_name='add-disk',
-                action_params={'osd-devices': loop_dev,
+                action_params={'osd-devices': osd_info[unit]['block-device'],
+                               'osd-ids': osd_info[unit]['osd-id'],
                                'partition-size': 5}
             )
             zaza_utils.assertActionRanOK(action_obj)
         zaza_model.wait_for_application_states()
 
-        logging.info('Removing previously added disks')
-        for param in params:
-            osd_id = self.get_local_osd_id(param['unit'])
-            param.update({'osd-id': osd_id})
-            action_obj = zaza_model.run_action(
-                unit_name=param['unit'],
-                action_name='remove-disk',
-                action_params={'osd-ids': osd_id, 'timeout': 5,
-                               'format': 'json', 'purge': False}
+        logging.info('Purge/Removing added OSDs')
+        for unit in list(osd_info):
```

Comment:
as above.

Reply:
Thanks for your comments @ajkavanagh. I have resolved them. Currently waiting for [PR#872856](https://review.opendev.org/c/openstack/charm-ceph-osd/+/872856) for verification.